### ISSUE TITLE
Your requests for V2.0.4 (and maybe for V2.1)

### ISSUE THREAD

**Issue by alvyxaz** [alvyxaz/barebones-masterserver#84]
As you've probably noticed from my absence in the communities - I'm a bit busy these days. 

However, I feel an urge to work on the framework for a bit, so I thought I'd ask - what do you want me to work on for the **V2.0.4** (minor release), and if you have any suggestions for bigger features - let me know what you'd like to see in **V2.1**.

For **V2.0.4** - request small, urgent changes, fixes and improvements
For **V2.1** - request something bigger, like a new module

Cheers!

**Comment by griezmann23** [comment]
shop, friend/community system, or maybe clan or guild system too :+1:

**Comment by arcanor** [comment]
In my opinion, there's nothing more important than improved documentation, stability of what's already provided, and maybe some benchmark testing. MSF already does plenty. I say let the community come up with the more specialized modules, while you focus on the core itself.

**Comment by recagonlei** [comment]
I guess for 2.1, you could focus on security, encripty connection, and so on... 
And for now, i though you could explain better about the profile modules... I did not understand very weel...  If you can explain how to use... I read the wiki, and looked at the demo scripts, and did not figure out how to do... And a question, the profiles can be used like a status, inventory, data about player and other stuff?

**Comment by angusmf** [comment]
What is hindering some people from being able to dig in and find answers in the code (and document them for the rest) is a lack of understanding of the general architecture and abstractions. Perhaps a guide on this, or review of the user docs that have covered the topic?

**Comment by RowellKataan** [comment]
Personally, I would love to see an in depth tutorial about the Profiles.  I tried writing some code using the current tutorial, but it works spotty at best between being used when the client is connected to the Master Server, and then connects to the Game Server.

**Comment by recagonlei** [comment]
It's kinda what is happening to me... I got some things, but i guess a improved docs about it or maybe a video tutorial, would be nice...

**Comment by arcanor** [comment]
I want to respond further about documentation... I think it's important that @alvyxaz is NOT the person writing every single document/tutorial/video. It would be best if people in the community stepped up and shared their experiences with MSF, as we struggle through the learning process, and finally win those battles. We should all try to share those tips and techniques that have worked for us.

For example, I've written a couple of wiki pages myself, referencing things I've learned and accomplished during my implementation process with MSF:

- https://github.com/alvyxaz/barebones-masterserver/wiki/RDBMS-Choices-for-Indie-Game-Developers:-PostgreSQL-vs.-MySQL
- https://github.com/alvyxaz/barebones-masterserver/wiki/Module-Mod---AuthModule-with-required-email-confirmation

In my opinion, @alvyxaz 's time is best spent doing things like considering important architecture decisions, doing code reviews for community pull requests, and occasional documentation of particularly tricky or new subjects that he's the only one who understands.

Basic tutorials can be done by the community, as we go through the different aspects of implementing MSF, and decide what's unclear... we can clear it up ourselves with new documentation, or additions to existing docs.

**Comment by recagonlei** [comment]
Thanks for that, dude... I undestand that, and i appreciate your contributions... I asked for help, just because i did not get to work properly, if somebody can help, it's welcome... And of course, i will contribute with what i will learn about msf, i started yesterday and i m studying properly...

**Comment by delmarle** [comment]
a complete example like the one that were discussed in a ticket: simple character creation, selection with only field like name, gender. include things like how to structure and what component to extend from, working scene provided.

 right now i feel v2 is much harder to dive in because of missing practical example

**Comment by TechieCrow** [comment]
I agree with @arcanor there is nothing more important then better documentation right now, I personally have struggled with it a little bit and hesitate to even try convert to MySQL (which is what I know and love) also still confused on profiles.

**Comment by arcticfire40** [comment]
Personally I would echo above posters that what is holding me up currently is my poor understanding of the over architecture. I would love to see a flow chart of sorts detailing how the classes are organized in the core, whats exposed, and how they are expected to be extended etc. I can and will if I must piece it together myself digging through the code but I always find it helpful to have the architect vision of the structure to reference. 

It would also greatly help those of more moderate coding skill or with limited time contribute if they knew which coding blocks should be added where to improve things.

**Comment by Raystorms** [comment]
Better documentation would be nice

if can, NAT punchthrough for 2.1

**Comment by HeadClot** [comment]
For 2.1 - I would like to see an [area of interest](http://gram.cs.mcgill.ca/papers/boulanger-06-comparing.pdf) module for much larger games using the MasterServer.

**Comment by Dracolique** [comment]
Well, given my recent issue that I opened [here](https://github.com/alvyxaz/barebones-masterserver/issues/97), I think compatibility with upcoming Unity .NET4.6 runtime is a fairly pressing issue.

**Comment by arcanor** [comment]
@HeadClot - I think AoI is something not relevant to MSF. It needs to be implemented as part of your custom game server.

**Comment by RikuTheFuffs** [comment]
Hi everybody, I think that a good thing would be a **clear guide on how the Auth Module can recognize already-registered user accounts within your database. I.E: for logging users in** . 

I'm asking this because most of pc-games developers will probably have their own websites where users register (and then download the game), and the current documentation is not really clear about how you should manage this kind of situation (where the users register from outside the game).

**I.E:** I have an independent **REST API system** that I can query to log my user in. Once I authenticate him through my API, which then writes the updated user status in the database and replies with a JSON to my login request, how can i tell my master server that this user is online so I can use his data for other msf things? (**I.E**: chat?) 

if what I wrote above sounds weird or nonsense to you, just think at my request as if it was "A detailed guide on how the Auth and Chat Modules work" :)

**P.S:** if you ever decide to allow the MSF to talk with REST APIs through common functions, I already have a couple of classes ready that I could share ^^

**Comment by jim-freeman** [comment]
@RikuTheFuffs Check out this issue: https://github.com/alvyxaz/barebones-masterserver/issues/83

It's a prereq for what you're talking about.

**Comment by RikuTheFuffs** [comment]
Hi @jim-freeman thanks for letting me notice that! I'll post my opinion about that in the Issue topic

**Comment by gremberman** [comment]
What about some voice chat?
It shall be great to have it.
We can use Speex C# codec for it - http://nspeex.codeplex.com

**Comment by PanMadzior** [comment]
It's always cool to see example modules, but in my opinion, they aren't major things. I would like to see some more security and architecture updates in the future. What I mean is that anyone who starts working on master servers should be already able to write some modules by their own.

Right now, the only additional module/feature I would like to see is the NAT Punchthrough. Since Photon Bolt was updated to 1.0 there is no more support for Zeus (additional program included with older versions of Photon Bolt which was giving possibilities to do a punchthrough for free). Now, they integrated Zeus with their Photon Cloud which is paid and CCU limited. Of corse, it isn't the feature for Bolt users only. It will help minimize server costs for anyone whose games can be hosted on players' local machines.

**Comment by optimisez** [comment]
I would like to see Redis database support in future release
